src/python_testing/TC_DeviceBasicComposition.py

```diff
@@ -941,6 +941,217 @@ def log_structured_data(start_tag: str, dump_string):
+
+        if len(listed_device_types) > 0:
+            if not all(x in {0x16, 0x0012, 0x0014} for x in listed_device_types):
+                self.fail_current_test()
+        self.print_step(3, "TH reads PartsList attribute- covered in step 2")
+
+        for endpoint_id, endpoint in self.endpoints.items():
+
+            self.print_step(8, "TH reads ServerList attribute for endpoint :{endpoint_id}")
+
+            if endpoint_id != 0:
+                if Clusters.Descriptor not in endpoint:
+                    continue
+
+                self.print_step(
+                    "1b", "TH reads DeviceTypeList and PartsList attributes from DUT for Endpoint {endpoint_id} supported by DUT (except Endpoint 0).")
+                parts_list_per_ep_non_zero = endpoint[Clusters.Descriptor][Clusters.Descriptor.Attributes.PartsList]
+
+                listed_device_types_ep_non_zero = [i.deviceType for i in endpoint
+                                                   [Clusters.Descriptor][Clusters.Descriptor.Attributes.DeviceTypeList]]
+                cluster_revision_ep_non_zero = endpoint[Clusters.Descriptor][Clusters.Descriptor.Attributes.ClusterRevision]
+
+                self.print_step(
+                    "1b.1", "Verify that the DeviceTypeList count is at least one for end point {endpoint_id}")
+
+                if len(listed_device_types_ep_non_zero) < 1:
+                    self.fail_current_test()
+
+                self.print_step(
+                    "1b.2", "If the DeviceTypeList contains more than one Application Device Type, verify that all the Application Device Types are part of the same superset for end point {endpoint_id}")
+                for item in listed_device_types_ep_non_zero:
+                    if item not in [14, 256, 19, 770, 17, 18, 20]:
+                        self.fail_current_test()
+
+                if cluster_revision_ep_non_zero <= 1:
+                    self.fail_current_test()
+
+                self.print_step(
+                    "1b.3", "Verify the DeviceTypeList does not contain the Root Node Device Type {endpoint_id}")
+                if ROOT_NODE_DEVICE_TYPE in listed_device_types_ep_non_zero:
+                    self.record_error(self.get_test_name(), location=AttributePathLocation(endpoint_id=0),
+                                      problem="Root node device type is listed on non zero endpoints", spec_location="Root node device type")
+                    self.fail_current_test()
+
+                self.print_step(
+                    "1b.4", "Revision should match the revision of that Device Type for endpoint {endpoint_id}")
+                device_revision_ep_non_zero = [i.revision for i in endpoint[Clusters.Descriptor]
+                                               [Clusters.Descriptor.Attributes.DeviceTypeList]]
+                for device_revision in device_revision_ep_non_zero:
+                    if device_revision != 1:
```

Comment:
replace all magic values with constants.

For specific revision values, we should read that from spec XML.